client/ping.go

```diff
@@ -13,18 +13,15 @@ func (cli *Client) Ping(ctx context.Context) (types.Ping, error) {
 		return ping, err
 	}
 	serverResp, err := cli.doRequest(ctx, req)
-	if err != nil {
-		return ping, err
-	}
-	defer ensureReaderClosed(serverResp)
+	if serverResp.header != nil {
```

Comment:
I was thinking we could split the `if serverResp.statusCode < 200 || serverResp.statusCode >= 400 {` block into a separate function that gets called after `doRequest`, if `doRequest` succeeds. The `API-Version` and `Docker-Experimental` header handling code could go in between. This wouldn't constitute a major refactoring, and it means the caller could immediately return an error if `doRequest` returns an error.